### DISCUSSION TITLE
Externalizing examples in the .spec file doesn't appear to allow linking to concrete response examples?

### DISCUSSION THREAD

**Discussion by markgargan** [specmatic/specmatic#748] [Ideas and Suggestions]
Hi folks, loving the framework! Am really hoping it's going to be a viable option for replacing our Spring Cloud Contract tests. I've an openapi yml spec for a quotation system we're building out. I can explicitly provide Request/Response pairs in the Examples section under the Schema, however, when I attempt to externalize these, it appears I can only externalize the requests. Providing Response Examples doesn't appear to be possible in the spec file in such a way that the requests produced in the Examples would marry up with the Response Examples. 

  Scenario Outline: Create Quote
    When POST /quot1
    Then status 200
    Then response
    Examples:
      | quoteReference        | customers |
      | VALID_QUOTE_REFERENCE | [{"type": "A", "dateOfBirth": "2023-01-01"}, {"type": "B", "dateOfBirth": "2022-01-01"}]|
    Response Examples:
       | fullPrice | monthlyPrice | status |
       |     10      |         20          |    OK   |

**Comment by joelrosario** [comment]
Thanks for raising this @markgargan. We'd love more participation on this topic. Moving to discussions.

**Reply by markgargan** [reply]
Sorry I think I replied to the issue rather than the discussion

Hi Joel, 

Yes this is almost working perfectly. I have a .spec file that references the openapi yaml. I've removed all Examples from the openapi spec file. I reference the .spec file as a contractsPath in the setup method of my ContractTests which extends the SpecmaticJunitSupport class. When I run it, it successfully swaps in the customerReference in the right place and it creates a single Customer in the 'customers' list and replaces the dateOfBirth value. However, the 'And response-body expectedResponse' doesn't check the response against the json object in the Examples column even though there's examples of this in the language.html file located at https://specmatic.in/documentation/language.html

I have the following table in the spec. 
```
  Scenario Outline: Valid Customer
    When POST /customer
    Then status 200
    And response-body expectedResponse
    Examples:
      | customerReference                    | dateOfBirth  | expectedResponse                                                         |
      | VALID_CUSTOMER_REFERENCE | 2025-01-01  | "{'fullPrice': '10.0', 'monthlyPrice': '20.0', 'customerStatus' : 'OK'}" |
```
A valid Customer POST looks like 
```
{
	"customerReference": "VALID_CUSTOMER_REFERENCE",
	"customers": [
		{
			"customerType": "A",
			"dateOfBirth": "2023-01-01"
		},
		{
			"customerType": "B",
			"dateOfBirth": "2022-01-01"
		}
	]
}```

**Reply by markgargan** [reply]
Ideally what I would like to do is to define a Customer Type with different customers that I could then reference in a customers column in the Examples table picking the customers to be in the array on the basis of the id in the table. Do you know if this is possible Joel? I found the gherking docs to be a little light on the ground wrt to the Examples sections 

```

  Scenario Outline: Valid Customer
    Given type Customer
      | id | customerType | dateOfBirth |
      | 1  | A            | 2021-01-02  |
      | 2  | B            | 2020-01-02  |
      | 3  | C            | 2019-01-02  |
      | 4  | D            | 2018-01-02  |
    When POST /customer
    Then status 200
    And response-body expectedResponse
    Examples:
      | customerReference                | customers                  | expectedResponse                                                         |
      | VALID_CUSTOMER_REFERENCE         | [Customer(1), Customer(2)] | "{'fullPrice': '10.0', 'monthlyPrice': '20.0', 'customerStatus' : 'OK'}" |
      | ANOTHER_VALID_CUSTOMER_REFERENCE | [Customer(3), Customer(4)] | "{'fullPrice': '20.0', 'monthlyPrice': '30.0', 'customerStatus' : 'OK'}" |
    
```

**Reply by joelrosario** [reply]
Specmatic's focus today is contract testing, which involves sending a specification-valid request to the application, and validating the signature of the response according to the specification.

Validation of specific scenarios like what's in your example (e.g. given X value in the request, the response should contain Y value) requires checking the values in the responses. This is a different kind of test as it's going beyond the response format. It's validating actual values and often also requires more types of assertions. Specmatic so far has left "non-contract" categories of tests to other test suites and testing tools.

That said, this question has come up more than once recently, and we are looking into it.

About the documentation, I realise that there's no explicit mention that only request values should go into rows, but not responses. Thanks for the feedback! If there's anything else you feel was missing when you were trying out the Gherkin wrapper feature, please let us know.

**Reply by markgargan** [reply]
Thanks very much for the detailed response Joel, I appreciate it! I believe half the problem is my lack of understanding of what the suite was attempting to achieve. Until recently I probably had the likes of integration/bdd and contract testing all bundled into the same head space but I'm starting to unravel the differences between them. In terms of what was missing from my point of view, I'm a Java developer working behind a corporate proxy, so I guess it's going to be mostly useful from that perspective
. Behind this corporate proxy, half the battle is working through it I'm afraid so a few of these notes'll be only specific to that. 

- Behind the corporate proxy, referencing the github repo in the specmatic.json wasn't working when going through the documentation. The ability to test against openapi yaml files local to the filesystem using the contractPaths system property should possibly be more prominent when running the ContractTest that extends the SpecmaticJunitSupport class.
https://specmatic.in/documentation/contract_tests.html#referring-to-local-specificatons

- Behind our corporate proxy, the location of the specmatic.jar is blocked. I believe it's because it's a jar rather than anything else. That jar itself, it appears to be a huge jar in comparison to the specmatic-core jar available in https://repo1.maven.org/. Would it be possible to release this up to the maven repo as well? I'd imagine most Java development houses like ours allow access to this repo. I tried building the source locally updating the necessary trustStore properties in the gradle.properties to allow access to the download 
https://services.gradle.org/distributions/gradle-7.6.1-bin.zip
,however, later in the build it's as if the trustStore was no longer being honoured? As if the gradle dependency resolution doesn't use the same gradle.properties file. I was getting issues like
 > Could not resolve org.jetbrains.kotlin:kotlin-gradle-plugin-api:1.7.21.
     Required by:
         project : > org.jetbrains.kotlin.plugin.serialization:org.jetbrains.kotlin.plugin.serialization.gradle.plugin:1.7.21 > org.jetbrains.kotlin:kotlin-serialization:1.7.21
         project : > org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.7.21 > org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.21
      > Could not resolve org.jetbrains.kotlin:kotlin-gradle-plugin-api:1.7.21.
         > Could not get resource 'https://plugins.gradle.org/m2/org/jetbrains/kotlin/kotlin-gradle-plugin-api/1.7.21/kotlin-gradle-plugin-api-1.7.21.pom'.
            > Could not GET 'https://jcenter.bintray.com/org/jetbrains/kotlin/kotlin-gradle-plugin-api/1.7.21/kotlin-gradle-plugin-api-1.7.21.pom'.
               > PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

- Providing the full specmatic.jar as a maven dependency would open the door for providing mvn goals/phases that run the command lines :-
    - specmatic stub
    - specmatic test

- The sample contract repo containing the api-order-v1.yaml file doesn't have a sample .spec file. You have to pull it together from the employees-without-examples.yml and the employee.spec in the documentation. 

- The maven dependency
```

<dependency>
      <groupId>org.springdoc</groupId>
      <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
      <version>${spring.doc.version}</version>
</dependency>

```
allows you to generate the equivalent openapi yaml spec file at
http://localhost:8080/v3/api-docs.yaml

I'm sure there's tools to generate the rest controllers/models from the openapi spec but in our case we're developing them manually.

Again very specific to a Spring-Boot Rest API Maven project but it would be nice to be able to have a git pre-hook that fires when a developer tries to commit that stands up the application in the same way that the ContractTest does, grabs the yaml specification file from the url above to a file say
src/test/contracts/openapi.new.yaml
and runs the yaml comparison 

java -jar specmatic.jar compare openapi.latest.yaml openapi.new.yaml 

If the openapi.new.yaml is no longer backward compatible, the build fails unless there's a 'force' flag of some kind provided in the case that the developer understands they're making a breaking change? 

If everything is okay then the file is renamed to openapi.latest.yaml and everything is then committed up to the repo.
I believe in that way the shift left goes all the way to the developers local machine before even reaching the repository

If this is useful and I can think of anything else I'll append to this chat.

**Reply by joelrosario** [reply]
Hi Mark, apologies for the delayed response. Thanks for sharing these notes. Please see our replies inline.

> Thanks very much for the detailed response Joel, I appreciate it! I believe half the problem is my lack of understanding of what the suite was attempting to achieve. Until recently I probably had the likes of integration/bdd and contract testing all bundled into the same head space but I'm starting to unravel the differences between them. In terms of what was missing from my point of view, I'm a Java developer working behind a corporate proxy, so I guess it's going to be mostly useful from that perspective . Behind this corporate proxy, half the battle is working through it I'm afraid so a few of these notes'll be only specific to that.
> 
> * Behind the corporate proxy, referencing the github repo in the specmatic.json wasn't working when going through the documentation. The ability to test against openapi yaml files local to the filesystem using the contractPaths system property should possibly be more prominent when running the ContractTest that extends the SpecmaticJunitSupport class.
>   https://specmatic.in/documentation/contract_tests.html#referring-to-local-specificatons

There have been several updates to Specmatic since the last message on this thread.
Specmatic recently added support for a filesystem spec provider, called `filesystem`, which reads files from specmatic.json. Here's the [documentation](https://specmatic.in/documentation/specmatic_json.html#use-specifications-on-local-file-system) for this.

However, thanks for the feedback on your experience being behind a company firewall. We've updated the section on [programmatically running contract tests](https://specmatic.in/documentation/contract_tests.html#programmatically-executing-specmatic-contract-as-tests) and on [specmatic.json configuration](https://specmatic.in/documentation/contract_tests.html#declaring-contracts-in-configuration) on the page on Contract Tests. Hope it helps.


> * Providing the full specmatic.jar as a maven dependency would open the door for providing mvn goals/phases that run the command lines :-
>   
>   * specmatic stub
>   * specmatic test

Specmatic's executable is now available directly from maven. You can learn more on the [download page](https://specmatic.in/download.html).

> * The sample contract repo containing the api-order-v1.yaml file doesn't have a sample .spec file. You have to pull it together from the employees-without-examples.yml and the employee.spec in the documentation.

We've introduced the ability to externalise test data to a JSON file, and are recommending this instead of Gherkin. You can read more about it [here](https://specmatic.in/documentation/contract_tests.html#externalising-examples--test-cases).

> * The maven dependency
> 
> ```
> 
> <dependency>
>       <groupId>org.springdoc</groupId>
>       <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
>       <version>${spring.doc.version}</version>
> </dependency>
> ```
> 
> allows you to generate the equivalent openapi yaml spec file at http://localhost:8080/v3/api-docs.yaml
> 
> I'm sure there's tools to generate the rest controllers/models from the openapi spec but in our case we're developing them manually.

Thanks for pointing out springdoc-openapi-starter-webmvc-ui. Note that when doing Contract Driven Development, the spec is created or updated first. The advantage of this that the API and consumer teams can collaborate to evolve or approve the API design before any code is written, and then proceed to develop their respective pieces independently (as I am guessing you are doing, from your comment). They would leverage contract tests and intelligent service virtualisation to stay in sync with the spec in parallel. API developers could use springdoc-openapi-starter-webmvc-ui as a convenient way to provide consumers with a way to explore the API.

> 
> Again very specific to a Spring-Boot Rest API Maven project but it would be nice to be able to have a git pre-hook that fires when a developer tries to commit that stands up the application in the same way that the ContractTest does, grabs the yaml specification file from the url above to a file say src/test/contracts/openapi.new.yaml and runs the yaml comparison
> 
> java -jar specmatic.jar compare openapi.latest.yaml openapi.new.yaml
> 
> If the openapi.new.yaml is no longer backward compatible, the build fails unless there's a 'force' flag of some kind provided in the case that the developer understands they're making a breaking change?
> 
> If everything is okay then the file is renamed to openapi.latest.yaml and everything is then committed up to the repo. I believe in that way the shift left goes all the way to the developers local machine before even reaching the repository
> 

Specmatic has a command named `compatible` that can analyse two git commits, such as HEAD and HEAD~1, in a pre-commit hook. It will return a non-zero value if a backward compatibility error is found. You can read more about it [here](). In general, for backward-breaking changes, we recommend changing the name of the file itself. However, I imagine one could use environment variables to pass a flag to the pre-commit hook so that it ignores errors, if needed.

Thanks again @markgargan for sharing these notes.

> If this is useful and I can think of anything else I'll append to this chat.
